[SPARK-20736][Python] PySpark StringIndexer supports StringOrderType

actuaryzhang opened the pull request:
## What changes were proposed in this pull request?
PySpark StringIndexer supports StringOrderType added in #17879.

Commit:
Python API to StringOrderType in StringIndexer

Commit:
fix typo

Commit:
fix typo

actuaryzhang commented:
@felixcheung

SparkQA commented:
**[Test build #76914 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76914/testReport)** for PR 17978 at commit [`e5c8dcf`](https://github.com/apache/spark/commit/e5c8dcfcdcb9fcb9586339c3efebe85670126fb6).

SparkQA commented:
**[Test build #76914 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76914/testReport)** for PR 17978 at commit [`e5c8dcf`](https://github.com/apache/spark/commit/e5c8dcfcdcb9fcb9586339c3efebe85670126fb6).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76914/
Test FAILed.

Commit:
fix style

SparkQA commented:
**[Test build #76916 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76916/testReport)** for PR 17978 at commit [`bd80b37`](https://github.com/apache/spark/commit/bd80b37d9728624c6455ceca12198ce763b32a91).

SparkQA commented:
**[Test build #76916 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76916/testReport)** for PR 17978 at commit [`bd80b37`](https://github.com/apache/spark/commit/bd80b37d9728624c6455ceca12198ce763b32a91).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76916/
Test FAILed.

Commit:
fix style

SparkQA commented:
**[Test build #76917 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76917/testReport)** for PR 17978 at commit [`1f336ab`](https://github.com/apache/spark/commit/1f336ab70719f4074f4ac69cc0bb4750723b0bd5).

SparkQA commented:
**[Test build #76917 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76917/testReport)** for PR 17978 at commit [`1f336ab`](https://github.com/apache/spark/commit/1f336ab70719f4074f4ac69cc0bb4750723b0bd5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76917/
Test PASSed.

actuaryzhang commented:
@viirya @MLnick @BryanCutler @yinxusen @brkyvz @HyukjinKwon @srowen 
Ping for reviews or comments. Thanks much.

HyukjinKwon commented on python/pyspark/ml/feature.py:
```diff
@@ -2115,22 +2115,32 @@ class StringIndexer(JavaEstimator, HasInputCol, HasOutputCol, HasHandleInvalid,
+                 stringOrderType="frequencyDesc"):
         """
-        __init__(self, inputCol=None, outputCol=None, handleInvalid="error")
+        __init__(self, inputCol=None, outputCol=None, handleInvalid="error", \
```
Probably, the leading `\` could be removed.

HyukjinKwon replied:
(Probably, the leading `\` could be removed.)

actuaryzhang replied:
@HyukjinKwon  Thank you. Added tests.

HyukjinKwon commented:
(I am not used to ML. I just left a trivial comment for Python.)

viirya commented:
Code changes looks good. But we need to add test for this.

viirya commented on python/pyspark/ml/feature.py:
```diff
@@ -2115,22 +2115,32 @@ class StringIndexer(JavaEstimator, HasInputCol, HasOutputCol, HasHandleInvalid,
+    stringOrderType = Param(Params._dummy(), "stringOrderType",
+                            "How to order labels of string column. The first label after " +
+                            "ordering is assigned an index of 0. Supported options: " +
+                            "frequencyDesc, frequencyAsc, alphabetDsec, alphabetAsc.",
```
alphabetDsec -> alphabetDesc

Commit:
add tests

SparkQA commented:
**[Test build #76928 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76928/testReport)** for PR 17978 at commit [`44f0a36`](https://github.com/apache/spark/commit/44f0a362dd085022de215e9ab8d9536145f20d4d).

SparkQA commented:
**[Test build #76928 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76928/testReport)** for PR 17978 at commit [`44f0a36`](https://github.com/apache/spark/commit/44f0a362dd085022de215e9ab8d9536145f20d4d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76928/
Test FAILed.

Commit:
fix test error

SparkQA commented:
**[Test build #76929 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76929/testReport)** for PR 17978 at commit [`f66a445`](https://github.com/apache/spark/commit/f66a4455aba7ffc69d1b397cb828879d84bb39a6).

SparkQA commented:
**[Test build #76929 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76929/testReport)** for PR 17978 at commit [`f66a445`](https://github.com/apache/spark/commit/f66a4455aba7ffc69d1b397cb828879d84bb39a6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76929/
Test PASSed.

actuaryzhang commented:
@viirya Thanks much for your review. I corrected the typo and added some tests.

actuaryzhang commented:
@felixcheung Could you take a look? Thanks.

felixcheung commented on python/pyspark/ml/feature.py:
```diff
@@ -2111,26 +2112,45 @@ class StringIndexer(JavaEstimator, HasInputCol, HasOutputCol, HasHandleInvalid,
+    stringOrderType = Param(Params._dummy(), "stringOrderType",
+                            "How to order labels of string column. The first label after " +
+                            "ordering is assigned an index of 0. Supported options: " +
+                            "frequencyDesc, frequencyAsc, alphabetDesc, alphabetAsc.",
```
I think this should be generated instead of hardcoded - you can find a example on python..

actuaryzhang replied:
@felixcheung `stringOrderType` is not a shared trait on the Scala side, and I thought only the shared traits should be automatically generated. 
I have looked at the code for other ML transformers, and many of them hard coded, for example, [Imputer](https://github.com/apache/spark/blob/master/python/pyspark/ml/feature.py#L933) and [OneHotEncoder](https://github.com/apache/spark/blob/master/python/pyspark/ml/feature.py#L1520). 
Please let me know if I'm wrong, and a reference to example would be greatly appreciated. Thanks much!

felixcheung replied:
hmm, ok, I see a few examples that they are not generated

holdenk replied:
I know were mixed on doing this, but I like including the default value in the docstring, makes the documentation closer to the Scala doc and makes it easier to read without having to refer to the ScalaDoc.

felixcheung commented on python/pyspark/ml/feature.py:
```diff
@@ -2082,8 +2082,9 @@ class StringIndexer(JavaEstimator, HasInputCol, HasOutputCol, HasHandleInvalid,
-    So the most frequent label gets index 0.
+    The indices are in [0, numLabels). By default, this is ordered by label frequencies
+    so the most frequent label gets index 0. The ordering behavior is controlled by
+    setting stringOrderType.
```
I think you need to backtick and add a tag for the attribute

actuaryzhang replied:
Added tag.

felixcheung commented on python/pyspark/ml/feature.py:
```diff
@@ -2139,6 +2159,20 @@ def setParams(self, inputCol=None, outputCol=None, handleInvalid="error"):
+    @since("2.3.0")
+    def getStringOrderType(self):
+        """
+        Gets the value of :py:attr:`stringOrderType` or its default value.
```
should it say what the default value is?

actuaryzhang replied:
OK, added default value.

Commit:
address comments

actuaryzhang commented:
@felixcheung Thanks so much for the review. I addressed most of the comments except auto generating code for defining `stringOrderType`. This parameter is not a shared trait on the Scala side, and I thought only the shared traits should be automatically generated. I have looked at the code for other ML transformers, and many of them hard coded, for example, [Imputer](https://github.com/apache/spark/blob/master/python/pyspark/ml/feature.py#L933) and [OneHotEncoder](https://github.com/apache/spark/blob/master/python/pyspark/ml/feature.py#L1520).

Please let me know if I'm wrong, and a reference to example would be greatly appreciated. 
Thanks much!

Commit:
minor style fix

SparkQA commented:
**[Test build #76967 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76967/testReport)** for PR 17978 at commit [`6acabc2`](https://github.com/apache/spark/commit/6acabc2f2d27cc25fd6cb52ff25c1ba2ce69bd23).

SparkQA commented:
**[Test build #76967 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76967/testReport)** for PR 17978 at commit [`6acabc2`](https://github.com/apache/spark/commit/6acabc2f2d27cc25fd6cb52ff25c1ba2ce69bd23).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76967/
Test PASSed.

actuaryzhang commented:
@viirya @felixcheung Any additional changes needed for this one?

felixcheung commented:
LGTM, ping @holdenk @jkbradley if they are interested

viirya commented:
LGTM

actuaryzhang commented:
@felixcheung Would you help merge this? Thanks.

felixcheung commented:
I'd hold this for another 3-4 days just in case..

MLnick commented:
LGTM @felixcheung

holdenk commented:
One minor optional comment, but not a blocker so LGTM (although if you decide to update the docstring LGTM pending tests).

Commit:
add default value for stringOrderType in docstring

actuaryzhang commented:
@holdenk  Thanks for the comment. Added default value in docstring. 
@felixcheung Please let me know if there is anything else needed for this PR. 
Thanks everyone for the review and comments!

SparkQA commented:
**[Test build #77131 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77131/testReport)** for PR 17978 at commit [`2fe9432`](https://github.com/apache/spark/commit/2fe9432945f16b77916244b0cc36ff07cdb53693).

SparkQA commented:
**[Test build #77131 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77131/testReport)** for PR 17978 at commit [`2fe9432`](https://github.com/apache/spark/commit/2fe9432945f16b77916244b0cc36ff07cdb53693).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77131/
Test FAILed.

Commit:
fix example error

SparkQA commented:
**[Test build #77132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77132/testReport)** for PR 17978 at commit [`5bfa4dc`](https://github.com/apache/spark/commit/5bfa4dc3ba60655d9a9ce4aded935303b90d33cb).

SparkQA commented:
**[Test build #77132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77132/testReport)** for PR 17978 at commit [`5bfa4dc`](https://github.com/apache/spark/commit/5bfa4dc3ba60655d9a9ce4aded935303b90d33cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77132/
Test PASSed.

felixcheung reviewed (approved):
LGTM

felixcheung commented:
merged to master. thanks everyone!
